MSC4075: MatrixRTC notifications & call ringing

toger5 opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixrtc-call-ringing/proposals/4075-rtc-notification-event.md)

To-do:
- [x] How do you decline a call? [MSC4310](https://github.com/matrix-org/matrix-spec-proposals/pull/4310)
- [x] Implementations using revised format from [`04635eb` (#4075)](https://github.com/matrix-org/matrix-spec-proposals/pull/4075/commits/04635ebc908e8299c77c0abe7c025cab393a6eed)
   - [x] Ruma https://github.com/ruma/ruma/pull/2199
   - [x] js-sdk https://github.com/matrix-org/matrix-js-sdk/pull/4826
- [x] Implementation using revised formats with relations [`7c996bb` (#4075)](https://github.com/matrix-org/matrix-spec-proposals/pull/4075/commits/7c996bb41f265f0313e31c091824c1f885fba5eb)
   - [x] js-sdk https://github.com/matrix-org/matrix-js-sdk/pull/4826
    - [x] Ruma https://github.com/ruma/ruma/pull/2199
- [x] Implementation using revised formats with lifetime [`6362c23` (#4075)](https://github.com/matrix-org/matrix-spec-proposals/pull/4075/commits/6362c2353bc06a7db25a513839928a75ea5c976c)
   - [x] js-sdk https://github.com/matrix-org/matrix-js-sdk/pull/4826
   - [x] Ruma https://github.com/ruma/ruma/pull/2199
- [x] This looks very similar to [m.call.invite](https://spec.matrix.org/v1.12/client-server-api/#mcallinvite) but with much less info. Have we taken into account learnings from this existing capability?
   - edit by @toger5: looking into this it seems that the `lifetime` concept makes sense. It allows the sending client to know/control when the receiver stops ringing. Additional precautions need to take place to not allow for infinite rings. This is part of [`6362c23` (#MSC4075)](https://github.com/matrix-org/matrix-spec-proposals/commit/6362c2353bc06a7db25a513839928a75ea5c976c)


Dependencies:

- https://github.com/matrix-org/matrix-spec-proposals/pull/4143

Implementations:
- Element Web
  - [x] sending rings: https://github.com/matrix-org/matrix-react-sdk/pull/11870
  - [x] ringing: https://github.com/matrix-org/matrix-react-sdk/pull/11870
- Ruma
  - [x] notify event support: https://github.com/ruma/ruma/pull/1704
- Element X
  -  [x] android sending rings
  -  [x] android ringing
  -  [x] ios sending rings
  -  [x] ios ringing
- Extensible events fallback
  - [ ] [Element web](https://github.com/element-hq/element-web/issues/28166)
  - [ ] Element X ios
  - [ ] Element X android

toger5 committed:
add initial msc for call ringing

Signed-off-by: Timo K <toger5@hotmail.de>

toger5 committed:
update msc number

Signed-off-by: Timo K <toger5@hotmail.de>

toger5 committed:
json->plaintext for example

Signed-off-by: Timo K <toger5@hotmail.de>

turt2live commented on proposals/4075-call-notify-event.md:
Please limit lines to approximately 120 characters.

toger5 committed:
markdown lint

Signed-off-by: Timo K <toger5@hotmail.de>

davidegirardi commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,142 @@
+notification to others. Since this will make clients ring this has a higher
+effect on the receiver. Since ringing has to obey the mute settings, it is
+very easy for the targeted users to mitigate the "attack". It can be very
+much compared to spamming a room with "@room" messages.
```
Can we add some words about default power levels here or somewhere else?

We could alternatively model a default behaviour in the spec itself so we avoid issue by client applications forgetting to implement default power levels.

toger5 replied:
I updated the msc. (thanks for the feedback)

davidegirardi commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,142 @@
+and similar to a notification banner.
+This is not enforced by the spec however and ultimately a client choice.
+
+Notify conditions depend on:\
```
I don't see anything about closed/completed calls. What should a client do if it was offline and it receives the state event when backfilling?

davidegirardi commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,142 @@
+  (It is possible to send a `m.call.notify` for an ongoing call if that
+  makes sense. Starting a call ahead of time, planning in a small group,
+  ringing another set of users at a specific time so they don't forget to join.
+  Ringing one specific user again who missed joining during the first ring.)
```
Sending an empty `"user_ids": []` array is also OK right?
If someone wants to start an optional "drop-in" call.

toger5 replied:
Starting a call is transparent via the call.mebers event anyways. So a notify event without any content (room= false and empty user_ids) is equivalent to not sending a notify event.

The client will display ongoing calls based on memberships anyways

toger5 committed:
review changes

Signed-off-by: Timo K <toger5@hotmail.de>

uhoreg commented:
@toger5 You've added some implementations in the PR description.  Do these implement the MSC in full, or are there parts of the MSC that are still lacking implementation?

toger5 commented:
Yes it's implemented in full. At least on the platforms shown in the description (EW).

There are optional configurations that are still missing. (Allowing to send a notification to a specific subset of users) but the API for this is already part of the SDK. There is no ui for this however and we are not sure we want that in EW.

TobiasFella commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,164 @@
+  start a new call. Because they would automatically send an event by joining.
+- There would be less traffic. With the proposed solution the first one who joins
+  needs to send a `m.call.notify` event and a `m.call.member` state event.
+- Very flexible, since the `m.call.notify` event is separate from a call
+  participation. This allows an external instance (a meeting organizer bot) to
+  just ring all the users which are invited to a meeting without needing to be
+  part of the call.
```
This sounds like an advantage of *separate* events, not of the *combined* approach

toger5 replied:
This is a mistake. Thanks for finding it.

toger5 committed:
review fixes

Signed-off-by: Timo K <toger5@hotmail.de>

giomfo commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,165 @@
+Legacy calls are using room events to negotiate the call.
+A client could use the initial steps in the negotiation to also make the phone ring.
+
+With matrix RTC based calls this signalling is done over state events.
```
@toger5 what are these state events? Are these all `m.call.member` events?

toger5 replied:
Yes exactly

giomfo commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,165 @@
+  If the user is not part of the `m.mentions` section as defined in
+  [MSC3952](https://github.com/matrix-org/matrix-spec-proposals/pull/3952)
+  this event can be ignored. (Push notifications are automatically filtered
+  so this only is important for events received via a sync)
```
@toger5 is this `m.call.notify` event encrypted when it is sent in an encrypted room? 
I think it is, so the Push notification could not be filtered server side. The notification will be filtered locally by the client. 
Note: If this event is encrypted, the call notification may have some trouble in encrypted mentions-only rooms, like it is described in
[MSC3996: Encrypted mentions-only rooms](https://github.com/matrix-org/matrix-spec-proposals/pull/3996). Indeed the encrypted event will not be pushed by the server in that case.

toger5 replied:
That is good to know. I am wondering if a @room notify event should be unencrypted in that case. This would sacrifice privacy but make it more reliable.
If this is transparent to the user that could be acceptable maybe.

toger5 replied:
Thanks for also linking MSC3996.

toger5 replied:
I added a `Limitations and recommendations` [section ](https://github.com/matrix-org/matrix-spec-proposals/blob/a57720b3b887214e84c94359aa75bdbb49db8eb5/proposals/4075-call-notify-event.md#L93-L115) so this is explicitly mentioned.

td-famedly commented on proposals/4075-call-notify-event.md:
```diff
@@ -0,0 +1,165 @@
+
+### Explicitly unspecified conditions
+
+- The duration of the ring sound is a deliberately chosen
```
Instead of hardcoding 10 seconds, I am assuming the clients could probably just look at the `lifetime` field for the invite event mapped to `call_id` from the notify event.

toger5 replied:
Which invite event are you referring to? The general idea is to allow the receiver to choose a ring duration and not the sender.
So reading from any event would result in someone else choosing how long you device should ring.

td-famedly replied:
> The general idea is to allow the receiver to choose a ring duration and not the sender.

To me that sounds a bit counter intuitive, f.ex I would not want to stop ringing the device after 10 seconds if the m.call.invite lifetime was 60 seconds and I can still pick up the call. Or does the "stop ringing after 10 seconds" also automatically send a reject then? (might be out of scope for this PR)

toger5 replied:
This is independent of the legacy call invite event. It is just what we use for matrix RTC sessions where there is no enforced call buildup procedure. (You just post your state event)
So most likely there is no call.invite event which lifetime could be used. The call notify event is the one that could include a lifetime. But we decided against that as described in the msc and the comment above.
It's also a slightly different semantic since the calling participant can stay in the call for however long they desire. Where a call invite has a Lifetime.

toger5 replied:
This is now referenced in the MCS: https://github.com/matrix-org/matrix-spec-proposals/commit/aee19b084886bf28b94aff93841565b41a3cb2c7

toger5 committed:
add Limitations and recommendations

Signed-off-by: Timo K <toger5@hotmail.de>

toger5 committed:
fixes

toger5 committed:
Add extensible event fallback

ara4n committed:
make ugly json highlighting errors go away

ara4n committed:
readability fixes

hughns committed:
Latest

toger5 committed:
rename `m.rtc.notify` -> `m.rtc.noitfication`

toger5 committed:
introduce relation (instead of copying the session data into the notify event)(

toger5 committed:
add `lifetime`

toger5 committed:
reference to rind duration discussion

toger5 committed:
add extensible events section

toger5 committed:
update relation strings

toger5 committed:
Iterating:
 - dont use custom rel_type
 - numerous spelling gramar issues and various more explicit explanations.

Signed-off-by: Timo K <toger5@hotmail.de>

BillCarsonFr commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+    "lifetime": 30000,
+    "m.mentions": {"user_ids": [], "room": true | false},
+    "m.relates_to": {"rel_type":"m.reference", "event_id":"$rtc_member_event_id"},
+    "notification_type": "ring | notification",
```
Looks like the current implementation in js-sdk uses `m.call.intent`https://github.com/matrix-org/matrix-js-sdk/blob/4d0d32307eb4f1ce1fb65080fcca704f5bdedc31/src/matrixrtc/types.ts#L150

BillCarsonFr replied:
Also in this MSC it is never talking about calls, it is talking about rtc sessions. So maybe the naming is bad?
Maybe it should just be a `rtc_intent` that depends on the actual session beeing notified? like `voice_call` | `video_call` | `white_board_invite` ...

Johennes replied:
> Also in this MSC it is never talking about calls, it is talking about rtc sessions. So maybe the naming is bad? Maybe it should just be a `rtc_intent` that depends on the actual session beeing notified? like `voice_call` | `video_call` | `white_board_invite` ...

Yeah, `m.call.intent` doesn't fit well with the generic MatrixRTC design of this proposal. Maybe we could keep this information in the `application` objects in slots or member events?

Johennes replied:
Actually, it is part of #4196 already.

Johennes replied:
This is now part of the notification event after ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+With [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
+this signalling is done over state events. Furthermore, [MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
+allows for large group calls (not just 1:1) which makes it very
+desirable to have more configurations over the ringing process.
```
I think this may still refer to the initial version of MatrixRTC from before sticky events.

```suggestion
With [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
this signalling is done over a mixture of state events and sticky
timeline events. Furthermore, [MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
allows for large group calls (not just 1:1) which makes it very
desirable to have more configurations over the ringing process.
```

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+- `lifetime` the relative time to the `sender_ts` for which the `ring` is active or to define the window in which the
+  `notification` is not ignored. The recommended value is **30 seconds**.
+  The receiving client **SHOULD** cap the lifetime to an upper bound. (recommended: **2 minutes**).
+  `lifetime` **MUST** be non-negative and SHOULD NOT exceed 120000 ms.
```
To avoid ambiguity, we could make the unit of time explicit here.

```suggestion
- `lifetime` the time in milliseconds relative to `sender_ts` for which the `ring` is active / the
  `notification` is not ignored. The recommended value is **30 seconds**.
  The receiving client **SHOULD** cap the lifetime to an upper bound. (recommended: **2 minutes**).
  `lifetime` **MUST** be non-negative and SHOULD NOT exceed 120000 ms.
```

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+  If available this can be used to gather session data for this notification event.
+  It should always be used if the notification is related to an RTC session.
+  (Eventually this event might be used to send
+  RTC session independent ringing notifications with a custom message. A bot-controlled tea timer for instance.)
```
For such a session-independent event, how would a receiving client know which slot to join when there is more than one in the room? I think this use case might need further thought and we should consider making the relation mandatory for the time being. That way receiving clients can route into the correct slot reliably.

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+The `session` information for this notification event can be obtained by the relation to the `m.rtc.member` event.
+
+In the following we define **call** as any MatrixRTC session with the
+same `"session"` contents.
```
What is `"session"` here? It doesn't seem to be defined anywhere.

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+and similar to a notification banner.
+This is not enforced by the spec however and ultimately a client choice.
+
+The client should only inform the user if all of the following conditions apply:
```
I think this formulation is slightly misleading because most of the conditions below are in the format "when X then DON'T notify".

```suggestion
The following rules apply to the receiving client when notifying the user:
```

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+  [Client-Server API](https://spec.matrix.org/v1.11/client-server-api/#definition-mmentions),\
+  the event should be ignored. (Push notifications are automatically filtered
+  so this only is important for events received via a sync)
+- Local notification settings:\
```
Is silencing a room really a local setting? Wouldn't you use push rules for this as well?

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+  room's state `m.rtc.member` event of the user contains a membership for
+  the **call** in the `m.rtc.notification` event.
+  This includes stopping the current ring sound if the room state updates so
+  this condition is true.
```
I think this needs updating to the new sticky events design.

```suggestion
  None of the devices should ring if they receive an `m.rtc.notification` event
  if there is any connected `m.rtc.member` event in the **call** associated
  with the `m.rtc.notification` event.
  This includes stopping the current ring sound if such a membership event
  appears.
```

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+  (could even happen for screen shares if the screen share is implemented as a
+  separate participating device) or MatrixRTC business logic of the call.)
+  This could result in a lot of push notifications with no obvious/simple way to
+  filter them.
```
I think in encrypted rooms this is true regardless, isn't it? Push Rules need to be re-evaluated after decryption on the client and that only works if the client gets push notifications for each encrypted event.

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+- `notification_type` required string:\
+  The type of notification to send.\
+  `ring`: The client should ring.\
+  `notification`: The client should show a notification.\
```
The `notification` use case is slightly unclear to me – especially why it is combined with a lifetime.

For ringing, the lifetime makes a lot of sense. You ring because you're waiting in the call.

But if you don't ring, is the use case not more of a singular notification in the sense of a duration-less "call started" ping? Why would you care about the duration the banner is shown to the recipient in this case?

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes commented on proposals/4075-rtc-notification-event.md:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+The fields are defined as follows:
````
#4196 adds further application-specific fields. We should maybe explicitly state here that the fields can be extended based on the MatrixRTC app?

Johennes replied:
Resolved with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes committed:
Update proposal to latest flavour of MatrixRTC and include outcome of internal discussions

Signed-off-by: Johannes Marbach <n0-0ne+github@mailbox.org>

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,280 @@
+  The local timestamp on the sending device when the event was emitted. This is used in combination
+  with `lifetime` to evaluate if the notification event is valid. To mitigate clients lying about
+  the `sender_ts`, this value has to be checked against `origin_server_ts`. Receivers SHOULD use
+  `origin_server_ts` if `|sender_ts - origin_server_ts| > 20000 ms`.
```
Should we clarify why we trust the sender's local time to be more accurate than the sending server's local time?

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+
+A new event `m.rtc.notification` is proposed which can be sent by a client that
+wants to notify others about the existence of a session for a MatrixRTC application.
+This event is added to the push rules for clients which
```
I think this has been made slightly more clear with ffaad72f8a5aa3377a8f77176f498f8f296ee86d.

Johennes committed:
Make required properties more explicit

Signed-off-by: Johannes Marbach <n0-0ne+github@mailbox.org>

Johennes commented on proposals/4075-rtc-notification-event.md:
```diff
@@ -0,0 +1,227 @@
+
+The client should only inform the user if all of the following conditions apply:
+
+- `m.rtc.notification` content:\
```
I have made the mentions required with ed9e4c3b01ec6c33b60c92bd37f84e666784c672.

Johennes committed:
Fix typo

Signed-off-by: Johannes Marbach <n0-0ne+github@mailbox.org>

Johennes commented on proposals/4075-rtc-notification-event.md:
@ara4n [said](https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r1856299594):

> Something which is very unclear right now is the lifecycle of a call (and the draft of MSC4196 doesn't really help).
>
> Specifically, what has happened to `m.call.invite`? `m.call.answer`? `m.call.select_answer`? `m.call.hangup`? The intention of MSC3401 was to keep using these (as per https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/group-voip/proposals/3401-group-voip.md#basic-call) even when an SFU is involved, in order to preserve signalling for:
>
> * Call was placed but isn't ringing yet (not that we have in legacy VoIP)
> * Call is ringing
> * Caller gave up and stopped ringing
> * Call has been answered on a device
> * Call was rejected locally
> * Call was rejected on all devices
> * Call was answered
> * Call was hung up
>
> i.e. all the state machine transitions that go into the lifecycle of a call.
>
> I'm very worried that this seems to have been collapsed to a single `m.call.notify` event from MSC4075 + `m.rtc.member` events here, which seems to be missing most of the above. For instance, I can't see how the caller tells the callee that it's no longer ringing them - not to mention stuff like early media (MSC3635) or ringback tones.
>
> Legacy VoIP was already a dumbed down version of SIP, and I'm worried that the lack of signalling semantics here is going to make mimicing SIP or PSTN semantics really hard (let alone bridging to it), and would be a backwards step from legacy voip.

Johennes replied:
@fkwp and @ara4n had a chat about this out of band and agreed to consider this resolved. The RTC member event lifetime and the RTC notification event should cover the required lifecycle.
